### PR TITLE
Add workflow for blue-green deployments

### LINKED ISSUE #77: Workflow for blue-green deployment while preserving original deployment name
Problem:
I want to define a *single* deployment and service in ksonnet/git. The resource names of the corresponding deployment and service would be jesse-deploy and jesse-svc, respectively. I will have 5 replicas of the deployment. I want the ability to perform a blue-green deployment where I surge to 10 concurrently running pods, half of which are running the newer version, then perform a DNS/service level cutover to the new version via an update of the jesse-svc service resource with the new label selector.

The challenge is that we want to preserve the deployment resource/name, while also not wanting to redefine the same deployment twice as two deployment components in the ksonnet app (i.e. blue and green). In other words, I do not want jesse-deploy-green.yaml and jesse-deploy-blue.yaml deployment manifests checked in anywhere in git.

Current blue-green kubernetes techniques describe the creation of a new deployment with a new name and updating the service to point to the new deployment. However, this does not work well with GitOps because it implies the new deployment and new name would need to be checked-in with a new name.

I am proposing the following workflow to achieve blue-green deployment while preserving the original deployment resource, and without requiring a user to manage two duplicate deployment manifests in their git repo.

1. a deployment jesse-deploy and service jesse-svc start out running version 1 of the image, with selector app=jesse and 5 replicas.
2. blue-green deploy action is invoked via argo-cd API
3. a new, ephemeral deployment jesse-deploy-tmpXYZ is created at version 1, with selector app=jesse-tmpXYZ
4. jesse-deploy-tmpXYZ becomes healthy with 5 replicas. At this point in time, we have 10 pods running version 1 of the image, split equally between the two deployments. But no traffic is directed to the ephemeral deployment jesse-deploy-tmpXYZ because the label is unique.
5. service jesse-svc is patched to use selector app=jesse-tmpXYZ (which is still running version 1). At this point, all traffic is directed to the ephemeral deployment jesse-tmpXYZ running version 1
6. jesse-deploy is updated with the new manifest to run version 2
7. jesse-deploy becomes healthy and with 5 replicas. At the point, we have 5 pods running version 2, and 5 pods running version 1. No traffic is being directed to version 2 yet.
8. The service jesse-service is updated with the original app selector app=jesse. At this point, all traffic is directed back to jesse-deploy which is running version 2.
9. jesse-deploy-tmpXYZ is no longer receiving any traffic. It is now safe to delete the ephemeral deployment. Once deleted, the live state of the deployment and service should now reflect the state in git and the app should be considered synced.

### PR DESCRIPTION
**Pull request by merenbach**
## Background

Currently Argo CD uses Kubernetes' default deployment strategy, which when updating a deployment does not bring up a parallel infrastructure prior to removing existing infrastructure.

This is intended to resolve #77.

## Implementation

Add a new Argo workflow to handle blue-green deployments with a minimum of external tools (just `jq`, for now).  **This is a work-in-progress.  Help is desired.**

## What's still broken or needs to be looked at?

1. Is the app suffix sufficient?  Just `-blue`?
2. How do we most _properly_ tell whether the service has cut over successfully once the app name is updated?  Are commands absolutely FIFO, so we don't really need to wait before issuing additional commands?
3. Should we be doing more to randomize the app name?
4. ~Are `{{GREEN_DEPLOYMENT_NAME}}`, `{{SERVICE_NAME}}`, and `{{MANIFEST_FILE}}` referenced properly so that Golang can substitute in variables?~
5. ~What command do we want for the actual application deployment upgrade?  I have a placeholder of `kubectl -o json apply -f {{inputs.parameters.manifest-file}} > /tmp/green-deployment.json"` right now.~
6. Where should this file be?  `examples/`?  Does it belong in the original Argo's `examples/` instead?
7. ~**This workflow completely untested.**   What additional legwork do I need to do (i.e., add some Go code somewhere) to be able to effectively test it?~

### CODE REVIEW

**Comment by jessesuen** [comment]
> This workflow completely untested. What additional legwork do I need to do (i.e., add some Go code somewhere) to be able to effectively test it?

You don't need any go code to test your workflow. You should only need `argo submit` to test your workflow. The workflow should be accept three input parameters:
1. service name
2. deployment name
3. new deployment manifest (as a json string)

Assuming you already have a running guestbook service/deployment at guestbook:v1. Then run your workflow manually (using argo submit) with the three inputs, including the new deployment spec containing the guestbook:v2 image tag. The workflow should be able to manipulate the existing service and deployment to follow the prescribed algorithm.

**Comment by merenbach** [comment]
@jessesuen can now invoke on a test nginx deployment with

    ./dist/argo submit bluegreen.yaml -p deployment-name=nginx-deployment -p service-name=nginx-service -p new-deployment-manifest="`cat nginx-service.yml`"

Getting this in response:

    2018/05/02 11:00:02 Workflow manifest bluegreen.yaml failed submission: the server could not find the requested resource (post workflows.argoproj.io)

**Comment by jessesuen** [comment]
This is great, and is pretty much what I was imagining.

> Is the app suffix sufficient? Just -blue?

I think thats fine, this is ephemeral so it shouldn't matter much

> How do we most properly tell whether the service has cut over successfully once the app name is updated? Are commands absolutely FIFO, so we don't really need to wait before issuing additional commands?

Patching the service with the label to do the cut over should be instantaneous, in my experience.

> Should we be doing more to randomize the app name?

If anything I prefer formulated vs. randomized because it makes resubmits/retries of the workflow predictable, and gives better opportunity to making the workflow steps idempotent.

> Where should this file be? examples/? Does it belong in the original Argo's examples/ instead?

Lets put this in a "workflows" directory for now. This blue-green is somewhat unusual because it has two service level cutovers, instead of one.

File: bluegreen.yaml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+          path: /tmp/green-deployment-app-name
+
+  - name: wait-deployment-ready
+    inputs:
+      parameters:
+      - name: deployment-name
+    container:
+      image: argoproj/argoexec:latest
+      command: [sh, -c]
+      args: ["kubectl rollout status --watch=true 'deployments/{{inputs.parameters.deployment-name}}'"]
```
Comment (jessesuen) [inline bluegreen.yaml]: Didn't know about this. Learned something new!

**Review by jessesuen — APPROVED** [review]
Great job! Let's create a workflows directory for this.

**Pull request merged**